src/github.com/matrix-org/dendrite/syncapi/sync/userstream.go

```diff
@@ -25,55 +27,108 @@ import (
 // goroutines can Broadcast(streamPosition) to other goroutines.
 type UserStream struct {
 	UserID string
-	// Because this is a Cond, we can notify all waiting goroutines so this works
-	// across devices for the same user. Protects pos.
-	cond *sync.Cond
+	// The waiting channels....... TODO
+	waitingChannels []chan<- types.StreamPosition
+	// The lock that protects pos
+	lock sync.Mutex
 	// The position to broadcast to callers of Wait().
 	pos types.StreamPosition
-	// The number of goroutines blocked on Wait() - used for testing and metrics
-	numWaiting int
+	// The time when waitingChannels was last non-empty
+	timeOfLastChannel time.Time
 }
 
 // NewUserStream creates a new user stream
-func NewUserStream(userID string) *UserStream {
+func NewUserStream(userID string, currPos types.StreamPosition) *UserStream {
 	return &UserStream{
-		UserID: userID,
-		cond:   sync.NewCond(&sync.Mutex{}),
+		UserID:            userID,
+		timeOfLastChannel: time.Now(),
+		pos:               currPos,
 	}
 }
 
-// Wait blocks until there is a new stream position for this user, which is then returned.
-// waitAtPos should be the position the stream thinks it should be waiting at.
-func (s *UserStream) Wait(waitAtPos types.StreamPosition) (pos types.StreamPosition) {
-	s.cond.L.Lock()
+// Wait returns a channel that produces a single stream position when
+// a new event *may* be available to return to the client.
+func (s *UserStream) Wait(ctx context.Context, waitAtPos types.StreamPosition) <-chan types.StreamPosition {
+	posChannel := make(chan types.StreamPosition, 1)
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
 	// Before we start blocking, we need to make sure that we didn't race with a call
 	// to Broadcast() between calling Wait() and actually sleeping. We check the last
 	// broadcast pos to see if it is newer than the pos we are meant to wait at. If it
 	// is newer, something has Broadcast to this stream more recently so return immediately.
 	if s.pos > waitAtPos {
-		pos = s.pos
-		s.cond.L.Unlock()
-		return
+		posChannel <- s.pos
+		close(posChannel)
+		return posChannel
 	}
-	s.numWaiting++
-	s.cond.Wait()
-	pos = s.pos
-	s.numWaiting--
-	s.cond.L.Unlock()
-	return
+
+	s.waitingChannels = append(s.waitingChannels, posChannel)
+
+	// We spawn off a goroutine that waits for the request to finish and removes the
+	// channel from waitingChannels
+	go func() {
+		<-ctx.Done()
```

Comment:
I don't suppose there's any elegant way to cancel this goroutine when the wait completes, so that we don't several piling up if there are are repeated calls in a single request?

Reply:
I guess another channel which is (just) closed when the wait completes, which you can select on here, would do. You might have to make `waitingChannels` a list of `(chan, chan)` tuples. Not sure if it's worth the additional complexity. WDYT?

Reply:
I guess we could actually have it so that the channels are kept alive so that a single sync requests only has one notification channel it reuses, and then we can close that channel when `ctx.Done()` returns. Still means an extra go routing per sync call, but is probably fine?

(Why are go channels so awful?)